Add noDuplicateGroupNamesInRegularExpressions lint rule

tatchi opened the pull request:
Add regexp lint rule to catch duplicated group names

sebmck commented on packages/@romejs/js-compiler/transforms/lint/noDuplicateGroupNamesInRegularExpressions.ts:
```diff
@@ -0,0 +1,38 @@
+            if (uniqueGroupNames.has(groupName)) {
+              context.addNodeDiagnostic(group, {
+                category: 'lint/noDuplicateGroupNamesInRegularExpressions',
+                message: `Duplicate group name <emphasis>${groupName}</emphasis> in regular expression`,
```
It'd be cool if we also highlighted the original occurence. These sorts of nice UX details will really set Rome apart from other tools.

To do it you can change `uniqueGroupNames` to a `Map<string, undefined | SourceLocation>` and then here do something like:

```
const advice = [];
const originalLoc = uniqueGroupNames.get(groupName);
if (originalLoc !== undefined) {
  advice.push({
    type: 'log',
    category: 'info',
    message: 'Originally defined at',
  });

  advice.push({
    type: 'frame',
    ...originalLoc,  
  });
}
```

And then put it into the diagnostic object.

sebmck replied:
The `| undefined` is because `node.loc` could be `undefined` if it was a node constructed dynamically without being created from source.

sebmck commented on packages/@romejs/js-compiler/transforms/lint/noDuplicateGroupNamesInRegularExpressions.ts:
```diff
@@ -0,0 +1,38 @@
+          const group = regExpBodyItem;
+          const groupName = group.name;
+          if (groupName !== undefined) {
+            if (uniqueGroupNames.has(groupName)) {
```
You can merge these conditions

sebmck commented on packages/@romejs/js-compiler/transforms/lint/noDuplicateGroupNamesInRegularExpressions.ts:
```diff
@@ -0,0 +1,38 @@
+    const {context, node} = path;
+
+    if (node.type === 'RegExpSubExpression') {
+      const uniqueGroupNames = new Set();
```
You should  use explicit type annotations on things like this since sometimes TS can get tripped up when doing inferrence.

tatchi commented:
Thanks for the review, I addressed your comments.

Note that if there are more than 2 occurrences of the same name, each error currently points out to the previously found location (`originalLoc`). See:

```js
 test.js:2:41 lint/noDuplicateGroupNamesInRegularExpressions ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Duplicate group name month in regular expression

    1 │ // let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<year>[0-9])-(?<day>[0-9])-([0-9])/;
  > 2 │ let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<day>[0-9])-([0-9])-(?<year>[0-9])-(?<month>[0-9])/;
      │                                          ^^^^^^^^^^^^^^^ 

  ℹ Originally defined here

    1 │ // let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<year>[0-9])-(?<day>[0-9])-([0-9])/;
  > 2 │ let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<day>[0-9])-([0-9])-(?<year>[0-9])-(?<month>[0-9])/;
      │           ^^^^^^^^^^^^^^^ 
```
```js
 test.js:2:94 lint/noDuplicateGroupNamesInRegularExpressions ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Duplicate group name month in regular expression

    1 │ // let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<year>[0-9])-(?<day>[0-9])-([0-9])/;
  > 2 │ let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<day>[0-9])-([0-9])-(?<year>[0-9])-(?<month>[0-9])/;
      │                                                                                               ^^^^^^^^^^^^^^ 

  ℹ Originally defined here

    1 │ // let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<year>[0-9])-(?<day>[0-9])-([0-9])/;
  > 2 │ let re = /(?<month>[0-9])-(?<year>[0-9])-(?<month>[0-9])-(?<day>[0-9])-([0-9])-(?<year>[0-9])-(?<month>[0-9])/;
      │                                          ^^^^^^^^^^^^^^^ 
```

I can make it so all the subsequent ones point out to the first defined group name. What do you prefer?

sebmck commented:
Up to you! What do you think is the most useful?

CaffeinateOften commented:
I sort of have this natural expectation that what we call the "original definition" of an instance of a thing (like a regex group name) would be the _first_ declaration that I read in the source code. This is the best/only rationale I could come up with for having the "originally defined here" portion point to the first instance of the regex group name.

sebmck commented:
An improved strategy is probably something like this:

```
if (node.type === 'RegExpSubExpression') {
  const groupUsage: Map<string, Array<RegExpGroupCapture>> = new Map();

  for (const bodyItem of node.body) {
    if (bodyItem.type === 'RegExpGroupCapture') {
      const groupName = group.name;

      if (bodyItem.group !== undefined) {
        let usages = groupUsage.get(bodyItem.group);
        if (usages === undefined) {
          usages = [];
          groupUsage.set(bodyItem.group, usages);
        }
        usages.push(bodyItem);
      }
    }

    for (const [name, usages] of groupUsage) {
      if (usages.size === 1) {
        continue;
      }

      const firstUsage = usages[0];

      const duplicateAdvice: PartialAdvice = usages.slice(1).map((node) => {
        if (node.loc === undefined) {
          return {
            type: 'log',
            category: 'warn',
            message: 'Unable to find location',
          };
        } else {
          return {
            type: 'frame',
            ...node.loc,
          };
        }
      });

      context.addNodeDiagnostic(firstUsage, {
        category: 'lint/noDuplicateGroupNamesInRegularExpressions',
        message: `Duplicate group name <emphasis>${groupName}</emphasis> in regular expression`,

        advice: [
          {
            type: 'log',
            category: 'info',
            message: 'Defined again here',
          },
          ...duplicateAdvice,
        ],
      });
    }
  }
}
```

So it would produce only one diagnostic per group name. This would have all the occurrences in the same diagnostic, and so would remove the cascade.

tatchi commented:
> An improved strategy is probably something like this:
> 
> ```
> if (node.type === 'RegExpSubExpression') {
>   const groupUsage: Map<string, Array<RegExpGroupCapture>> = new Map();
> 
>   for (const bodyItem of node.body) {
>     if (bodyItem.type === 'RegExpGroupCapture') {
>       const groupName = group.name;
> 
>       if (bodyItem.group !== undefined) {
>         let usages = groupUsage.get(bodyItem.group);
>         if (usages === undefined) {
>           usages = [];
>           groupUsage.set(bodyItem.group, usages);
>         }
>         usages.push(bodyItem);
>       }
>     }
> 
>     for (const [name, usages] of groupUsage) {
>       if (usages.size === 1) {
>         continue;
>       }
> 
>       const firstUsage = usages[0];
> 
>       const duplicateAdvice: PartialAdvice = usages.slice(1).map((node) => {
>         if (node.loc === undefined) {
>           return {
>             type: 'log',
>             category: 'warn',
>             message: 'Unable to find location',
>           };
>         } else {
>           return {
>             type: 'frame',
>             ...node.loc,
>           };
>         }
>       });
> 
>       context.addNodeDiagnostic(firstUsage, {
>         category: 'lint/noDuplicateGroupNamesInRegularExpressions',
>         message: `Duplicate group name <emphasis>${groupName}</emphasis> in regular expression`,
> 
>         advice: [
>           {
>             type: 'log',
>             category: 'info',
>             message: 'Defined again here',
>           },
>           ...duplicateAdvice,
>         ],
>       });
>     }
>   }
> }
> ```
> 
> So it would produce only one diagnostic per group name. This would have all the occurrences in the same diagnostic, and so would remove the cascade.

Brilliant! I've made the changes and that's way better now :)

Shouldn't we change the other `noDuplicate` rules to adopt the same strategy?

CaffeinateOften commented:
Hm, I think possibly! It might be case specific but I'd imagine the same structure of diagnostic information might be useful- I think the cognitive process that a developer goes through is the same. The first instance of the thing that is duplicated is important because it is the first - the rest are the duplicates that break the noDuplicate rule -- I can't come up with a reason that the 'originally defined here` info should be repeated again and again.

I was bias'ing my thoughts towards a terminal interface, but I feel like even the UI of an IDE showing this diagnostic information could take advantage of accessing the diagnostic information this way. I may have major gaps in understanding how ```@rome/diagnostics``` types are _meant_ to be used, but I imagine if the IDE UI part of this needed additional diagnostic information it would be added _when_ that need arises?

sebmck commented:
> Shouldn't we change the other `noDuplicate` rules to adopt the same strategy?

Sure! There might also be an opportunity to add a util method somewhere to avoid repeating this same pattern everywhere and have the diagnostic production in a single place.

sebmck commented:
Thank you so much!
